File: 9M4W/ehowlqk/boj_5639.py
import sys

input = sys.stdin.readline
sys.setrecursionlimit(10**9)

arr = []

while True:
    try:
        arr.append(int(input()))
    except:
        break


def gndnl(start, end):
    if start >= end:
        return
    
    mid = end
    for i in range(start + 1, end):
        if arr[i] > arr[start]:
            mid = i
            break
    
    gndnl(start + 1, mid)
    gndnl(mid, end)
    print(arr[start])

    return

gndnl(0, len(arr))

File: 9M4W/ehowlqk/boj_5639.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "boj_5639.txt";
const input = fs
  .readFileSync(filePath)
  .toString()
  .trim()
  .split("\n")
  .map((e) => Number(e));

function gndnl(start, end) {
  if (start >= end) return;
  let mid = end;
  for (let i = start; i < end; i++) {
    if (input[i] > input[start]) {
      mid = i;
      break;
    }
  }

  gndnl(start + 1, mid);
  gndnl(mid, end);
  console.log(input[start]);
}

gndnl(0, input.length);
